materialize-kafka/src/transactor.rs

```diff
@@ -0,0 +1,211 @@
+    let Store {
+        binding,
+        key_packed,
+        values_packed,
+        doc_json,
+        delete,
+        ..
+    } = store;
+
+    let b = &bindings[binding as usize];
+
+    let converted_key = convert_tuple(&key_packed);
+    let mut converted_values = if !values_packed.is_empty() {
+        convert_tuple(&values_packed)
+    } else {
+        vec![]
+    };
+    if !doc_json.is_empty() {
+        // The root document field, if selected, is always materialized as a
+        // string.
+        converted_values.push(serde_json::Value::String(doc_json));
+    }
+
+    Ok(match message_format {
+        MessageFormat::JSON => {
+            let key_doc = converted_key
+                .clone()
+                .into_iter()
+                .enumerate()
+                .map(|(i, value)| (b.field_names[i].clone(), value))
+                .collect::<serde_json::Value>();
+
+            let doc = converted_key
+                .into_iter()
+                .chain(converted_values)
+                .enumerate()
+                .map(|(i, value)| (b.field_names[i].clone(), value))
+                .collect::<serde_json::Value>();
+
+            serde_json::to_writer(&mut *key_buf, &key_doc)?;
+            let mut rec = BaseRecord::to(&b.topic).key(key_buf);
+
+            if !delete {
+                serde_json::to_writer(&mut *payload_buf, &doc)?;
+                rec = rec.payload(payload_buf);
+            }
+
+            rec
+        }
+        MessageFormat::Avro => {
+            let doc = converted_key
+                .into_iter()
+                .chain(converted_values)
+                .enumerate()
+                .map(|(i, value)| (b.field_names[i].clone(), value))
+                .collect::<serde_json::Value>();
+
+            let schema = b
+                .schema
+                .as_ref()
+                .expect("schema must be available for Avro encoding");
+
+            key_buf.push(0);
+            key_buf.extend(schema.key_schema_id.to_be_bytes());
+            encode_key(key_buf, &schema.key_schema, &doc, &b.key_ptr).context("encoding key")?;
+            let mut rec = BaseRecord::to(&b.topic).key(key_buf);
+
+            if !delete {
+                payload_buf.push(0);
+                payload_buf.extend(schema.schema_id.to_be_bytes());
+                encode(payload_buf, &schema.schema, &doc).context("encoding payload")?;
+                rec = rec.payload(payload_buf);
+            }
+
+            rec
+        }
+    })
+}
+
+fn convert_tuple(packed: &Bytes) -> Vec<serde_json::Value> {
+    tuple::unpack::<Vec<_>>(packed)
+        .expect("tuple must unpack")
+        .iter()
+        .map(|element| match element {
+            Element::Bool(v) => serde_json::json!(v),
+            Element::Int(v) => serde_json::json!(v),
+            Element::Float(v) => serde_json::json!(v),
+            Element::Double(v) => serde_json::json!(v),
+            Element::String(v) => serde_json::json!(v),
```

Comment:
nit: this could be rewritten like so

```suggestion
            Element::Bool(v) |
            Element::Int(v) |
            Element::Float(v) |
            Element::Double(v) |
            Element::String(v) => serde_json::json!(v),
```

Reply:
Hmm when I tried this it didn't work - I think the `v` would all have to be the same type, but they are different according to their enum variant, for example `bool`, `i64`, `f32`, etc.